File: frontend/src/api/templates.ts
/**
 * Workout templates API helpers.
 * Templates can also be applied into a session-shaped prefill payload for Today's Log.
 */
import { apiClient } from "./client";

const TEMPLATES_PREFIX = "/api/v1/templates";

export interface TemplateExercise {
  exercise_id: number;
  exercise_name: string;
  target_sets: number;
  target_reps: number;
}

export interface WorkoutTemplate {
  id: number;
  name: string;
  created_at?: string;
  updated_at?: string;
  exercises: TemplateExercise[];
}

export interface TemplateExerciseInput {
  exercise_id: number;
  target_sets: number;
  target_reps: number;
}

export interface TemplateCreateInput {
  name: string;
  exercises: TemplateExerciseInput[];
}

export interface TemplateUpdateInput {
  name?: string;
  exercises?: TemplateExerciseInput[];
}

export interface TemplateApplySet {
  set_number: number;
  reps: number;
  weight: number | string;
}

export interface TemplateApplyExercise {
  exercise_id: number;
  exercise_name: string;
  sets: TemplateApplySet[];
  notes?: string | null;
}

export interface TemplateApplyOut {
  template_id: number;
  template_name: string;
  exercises: TemplateApplyExercise[];
}

export async function listTemplates(): Promise<WorkoutTemplate[]> {
  const { data } = await apiClient.get<WorkoutTemplate[]>(`${TEMPLATES_PREFIX}/`);
  return data;
}

export async function getTemplate(id: number): Promise<WorkoutTemplate> {
  const { data } = await apiClient.get<WorkoutTemplate>(`${TEMPLATES_PREFIX}/${id}`);
  return data;
}

export async function createTemplate(input: TemplateCreateInput): Promise<WorkoutTemplate> {
  const { data } = await apiClient.post<WorkoutTemplate>(`${TEMPLATES_PREFIX}/`, input);
  return data;
}

export async function updateTemplate(id: number, input: TemplateUpdateInput): Promise<WorkoutTemplate> {
  const { data } = await apiClient.put<WorkoutTemplate>(`${TEMPLATES_PREFIX}/${id}`, input);
  return data;
}

export async function deleteTemplate(id: number): Promise<void> {
  await apiClient.delete(`${TEMPLATES_PREFIX}/${id}`);
}

export async function applyTemplate(id: number): Promise<TemplateApplyOut> {
  const { data } = await apiClient.get<TemplateApplyOut>(`${TEMPLATES_PREFIX}/${id}/apply`);
  return data;
}

File: backend/app/schemas/templates.py
"""
templates.py

Pydantic schemas for WorkoutTemplate CRUD and template-apply prefill payloads.
"""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import List

from pydantic import BaseModel, Field, field_validator


class TemplateExerciseIn(BaseModel):
    """Exercise row inside a template request."""

    exercise_id: int = Field(..., gt=0)
    target_sets: int = Field(..., gt=0)
    target_reps: int = Field(..., gt=0)


class TemplateExerciseOut(BaseModel):
    """Exercise row as returned in template responses."""

    exercise_id: int
    exercise_name: str
    target_sets: int
    target_reps: int


class TemplateCreate(BaseModel):
    """Request body for POST /templates."""

    name: str = Field(..., min_length=1, max_length=255)
    exercises: List[TemplateExerciseIn] = Field(default_factory=list)

    @field_validator("name")
    @classmethod
    def normalize_name(cls, a_value: str) -> str:
        """Trim the template name and reject blank values."""
        name = a_value.strip()
        if not name:
            raise ValueError("Template name is required")
        return name


class TemplateUpdate(BaseModel):
    """Request body for PUT /templates/{template_id}."""

    name: str | None = Field(default=None, min_length=1, max_length=255)
    exercises: List[TemplateExerciseIn] | None = None

    @field_validator("name")
    @classmethod
    def normalize_name(cls, a_value: str | None) -> str | None:
        """Trim the template name and reject blank values."""
        if a_value is None:
            return None
        name = a_value.strip()
        if not name:
            raise ValueError("Template name is required")
        return name


class WorkoutTemplateOut(BaseModel):
    """Template as returned by list/get/create/update."""

    id: int
    name: str
    created_at: datetime | None = None
    updated_at: datetime | None = None
    exercises: List[TemplateExerciseOut] = Field(default_factory=list)


class TemplateApplySetOut(BaseModel):
    """One placeholder set returned when applying a template."""

    set_number: int
    reps: int
    weight: Decimal = Decimal("0")


class TemplateApplyExerciseOut(BaseModel):
    """Exercise with placeholder sets returned by the apply endpoint."""

    exercise_id: int
    exercise_name: str
    sets: List[TemplateApplySetOut] = Field(default_factory=list)
    notes: str | None = None


class TemplateApplyOut(BaseModel):
    """Prefill payload returned by GET /templates/{template_id}/apply."""

    template_id: int
    template_name: str
    exercises: List[TemplateApplyExerciseOut] = Field(default_factory=list)
